src/catchup/CatchupWork.cpp

```diff
@@ -3,55 +3,230 @@
 
 #include "catchup/CatchupWork.h"
+#include "catchup/CatchupRange.h"
+#include "catchup/DownloadAndApplyBucketsWork.h"
+#include "catchup/DownloadAndApplyTransactionsWork.h"
+#include "catchup/DownloadAndVerifyLedgersWork.h"
+#include "catchup/ResolvedCatchupConfiguration.h"
 #include "history/HistoryManager.h"
 #include "historywork/GetHistoryArchiveStateWork.h"
-#include "lib/util/format.h"
+#include "ledger/LedgerManager.h"
 #include "main/Application.h"
+#include "util/Logging.h"
+#include <lib/util/format.h>
 
 namespace stellar
 {
 
 CatchupWork::CatchupWork(Application& app, WorkParent& parent,
-                         uint32_t initLedger, std::string const& mode,
-                         bool manualCatchup)
+                         CatchupConfiguration catchupConfiguration,
+                         VerifyLedgerMode verifyMode,
+                         ProgressHandler progressHandler)
     : BucketDownloadWork(
-          app, parent, fmt::format("catchup-{:s}-{:08x}", mode, initLedger),
+          app, parent, "catchup",
           app.getHistoryManager().getLastClosedHistoryArchiveState())
-    , mInitLedger(initLedger)
-    , mManualCatchup(manualCatchup)
+    , mCatchupConfiguration{catchupConfiguration}
+    , mVerifyMode{verifyMode}
+    , mProgressHandler{progressHandler}
 {
 }
 
-uint32_t
-CatchupWork::nextLedger() const
+std::string
+CatchupWork::getStatus() const
 {
-    return mManualCatchup
-               ? mInitLedger
-               : mApp.getHistoryManager().nextCheckpointLedger(mInitLedger);
+    if (mState == WORK_PENDING)
+    {
+        if (mDownloadAndApplyTransactionsWork)
+        {
+            return mDownloadAndApplyTransactionsWork->getStatus();
+        }
+        else if (mDownloadAndApplyBucketsWork)
+        {
+            return mDownloadAndApplyBucketsWork->getStatus();
+        }
+        else if (mDownloadAndVerifyLedgersWork)
+        {
+            return mDownloadAndVerifyLedgersWork->getStatus();
+        }
+        else if (mGetBucketsHistoryArchiveStateWork)
+        {
+            return mGetBucketsHistoryArchiveStateWork->getStatus();
+        }
+        else if (mGetHistoryArchiveStateWork)
+        {
+            return mGetHistoryArchiveStateWork->getStatus();
+        }
+    }
+    return BucketDownloadWork::getStatus();
 }
 
-uint32_t
-CatchupWork::archiveStateSeq() const
+void
+CatchupWork::onReset()
 {
-    return firstCheckpointSeq();
+    auto toLedger = mCatchupConfiguration.toLedger() == 0
+                        ? "CURRENT"
+                        : std::to_string(mCatchupConfiguration.toLedger());
+    CLOG(INFO, "History") << "Starting catchup with configuration:\n"
+                          << "  lastClosedLedger: "
+                          << mCatchupConfiguration.lastClosedLedger() << "\n"
+                          << "  toLedger: " << toLedger << "\n"
+                          << "  count: " << mCatchupConfiguration.count();
+
+    auto toCheckpoint =
```

Comment:
this seems wrong: the (complex) logic to compute what to download etc is in `makeCatchupRange` but that code is only triggered after kicking off `GetHistoryArchiveStateWork`

Reply:
oh nm, this is to fetch the remote state from archive